tests/Doctrine/Tests/DBAL/BugScenarioTest.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\DBAL;
+
+use Doctrine\DBAL\Schema\Comparator;
+use Doctrine\DBAL\Types\Type;
+use Doctrine\DBAL\Types\Types;
+use Doctrine\Tests\DbalFunctionalTestCase;
+
+final class BugScenarioTest extends DbalFunctionalTestCase
+{
+    public function setUp() : void
+    {
+        parent::setUp();
+
+        $schemaManager = $this->connection->getSchemaManager();
+        $schema        = $schemaManager->createSchema();
+
+        $originalTable = $schema->createTable('to_be_altered');
+        $column        = $originalTable->addColumn('some_column_name', Types::TEXT);
+        $column->setNotnull(false);
+
+        $schemaManager->createTable($originalTable);
+    }
+
+    public function testInvalidQueryIsGenerated() : void
+    {
+        $schemaManager = $this->connection->getSchemaManager();
+        $schema        = $schemaManager->createSchema();
+
+        $originalTable = $schema->getTable('to_be_altered');
+        $modifiedTable = clone $originalTable;
+
+        $column = $modifiedTable->getColumn('some_column_name');
+        $column->setType(Type::getType(Types::INTEGER));
```

Comment:
IMO this is where the existing design falls apart. It's like `$car->setType(Book::class)`.  It works in some cases and doesn't in others. There's just one more case when it doesn't work.

> I'd like to fix this, but I'm a little lost on where to do that. For example we could check the platform options when setting a type to the column and remove the irrelevant ones, but that would require specific logic for specific platform options in the column class.

Here's another example:
```php
$column = $table->addColumn('test', Types::TEXT, ['length' => 128]);

$column->setType(Type::getType(Types::INTEGER));
```
What's the expected column type and length given that both types support the "length" option?

Reply:
@morozov I totally agree, it would nicer to have another layer making `Column` just an abstraction. A column should be a type, not specify one.

So we could have an `IntegerColumn` or a `TextColumn` and have to transform it, if we want to change it using a `IntegerToTextColumnTransformer` of `TextToIntegerColumnTransformer` (Maybe all can available through a `ColumnTypeModifier` facade/helper for convience). Do you like that idea?

I guess a change like that would require a major version release though.

Reply:
> A column should be a type, not specify one.

I'd start with documenting the existing architectural layers, what properties they have, what problems they solve and how they should be regrouped in order to solve this problem as well. E.g.:

1. We have columns which can have types and attributes that are currently applied to any type. Columns are used for DDL.
2. We have types which are used to defined columns and to bind parameters to prepared statements.
3. Some column attributes seem to belong to a type in the DDL context (e.g. string length). But others belong to a column _of_ a given type (e.g. charset that is not part of the type but is a property of a column of type string).

I think it's too premature to discuss the solution of the problem w/o the problem statement and the existing design laid out.

Reply:
Alright I think you described the problem pretty well already, but I can dig into how `Column` is used exactly within `DBAL` and create some sort of report/proposal. It will take me some time though.